### DISCUSSION TITLE
`$ref` in generated JSON

### DISCUSSION THREAD

**Discussion by azzazzel** [xtclang/xvm#86] [General]
Hello folks,

I have a simple model like this
```
const Info(String name, String[] lines);
```
and an endpoint 
```
@Get("infos")
Info[] getInfos() {
   ...  
}
```

The endpoint returns an array containing few entries, each with a `name` and empty `lines`, and when I access the endpoint, I get 

```
[{
    "name": "name 1",
    "lines": []
}, {
    "name": "name 2",
    "lines": {
        "$ref": "/0/lines"
    }
}, {
    "name": "name 3",
    "lines": {
        "$ref": "/0/lines"
    }
}]
```

Now, I know that, strictly speaking, this is a correct JSON, but it's a pain to resolve the `$ref` on the client side. And in this particular case, it's not really an optimization, is it? Is there any way to disable the usage of `$ref`? I'm new to the language and the platform and I couldn't find a way in the limited time I have to explore it.

**Comment by cpurdy** [comment]
Yes, it's easy to get rid of the `$ref`'s ... see Schema.x in lib_json. The constructor has a bunch of options, including `enablePointers`, which is what causes the `$ref`'s to be added.

@ggleyzer How would he change the default options on the Schema that is used in his web app?

**Reply by ggleyzer** [reply]
The way it's done today, you would have to create a new schema instead of using the DEFAULT, for example:

      Schema mySchema = new Schema(
            enableReflection = True,
            enableMetadata   = True,
            enablePointers   = False,
            randomAccess     = True);

**Reply by cpurdy** [reply]
And how does the web-app pick up the schema? Does it reference DEFAULT directly, for example? How would we override it on an app by app basis?

**Comment by ggleyzer** [comment]
Since MyApp is WebApp,

    MyApp.registry.jsonSchema = mySchema;

**Comment by azzazzel** [comment]
Nice! Thank you!

It didn't work at first but after some digging, I figured it's actually called `registry_` (with an underscore at the end).

As I'm playing around with the platform, I added 
```
        this.registry_.jsonSchema = new Schema(
                enableReflection = True,
                enableMetadata   = True,
                enablePointers   = False,
                randomAccess     = True);
```
at the end of the `configure` method of the `platformUI.x` ([here](https://github.com/xtclang/platform/blob/d3cfcae67fd7ffae63694e1e12a61d33218c45dc/platformUI/src/main/x/platformUI.x#L34)). 
I'm not sure if that's the right way to do it but it worked ;)

**Reply by ggleyzer** [reply]
Sorry, my mistake. Since WebApp is a mixin into a module we didn't want to "take" good names from their apps - hence the underscore

**Reply by ggleyzer** [reply]
I don't think "configure" is the right place to change settings for *all* applications. Let me think what would be a good place to configure custom settings on a module; we may need to add some hook...

**Reply by azzazzel** [reply]
Just to clarify, my intention wasn't to change the settings for _all_ applications but to change the settings for the `platformUI` itself. Following our discussion at JCrete, I'm playing around with a new UI for it, and I needed to add some endpoints to it. 

Are you saying that all deployed applications inherit the registry and thus, my change will be applied to all? I'm sorry if my questions don't make sense. I'm still wrapping my head around this.

**Reply by cpurdy** [reply]
Your question makes perfect sense! Each application would have its own "registry", specific to that application.

**Reply by ggleyzer** [reply]
Oh.. you're right, it will not affect any other apps. I misunderstood your intent.

**Reply by cpurdy** [reply]
@azzazzel Any chance we can do a zoom call with you to see the approach you're taking and get feedback from you on what we can be improving?

**Reply by azzazzel** [reply]
Sure thing, @cpurdy! 
Actually, that was kind of my plan from the start - learn by making an MVPoC and then show it to you to see what you think. 
Let's plan something for next week if that work for you. Hopefully, by that time, I'll have something to show.
